[[ LCB ]] Implement lexical scoping of variables in blocks

runrevmark opened the pull request:
This patch makes variable definitions lexically scoped within blocks. e.g. The following won't compile as tVar is declared inside the repeat block:

```
   repeat forever
     variable tVar
   end repeat
   put tVar
```

In order to implement this a couple of VM changes have been made, as well as compiler changes.

The first change is that the VM internally now uses the nil pointer to represent an unassigned slot - this solves the previous problem of it not being able to tell the difference between a slot with 'optional' type and a slot which has not actually had a value assigned to it.

The second change is that giving variables default values is now the responsibility of the compiler. It now does this using a new opcode 'Reset' which resets a slot to its type based default, or to unassigned if the slot is untyped/its type has no default value.

runrevmark commented:
Test wise - I've moved the default values test to compiler rather than vm. Tests are needed for the new nil means unassigned and for the reset opcode, however that can't be done until 'bytecode blocks' are updated.

runrevmark commented:
There's some further work needed here with regards variable initialization - module level slots need to be initialized in the same way as local slots (i.e. via code generated by the compiler) but to do this modules need to gain an implicit 'init' and 'finit' method (compiler generated).

runrevmark commented:
Another default value related thing. At the moment, variables are defined at the point of declaration within a handler, and obviously in and inout parameters already have a value. The only part left for consideration is out variables - these currently need to be assigned to at some point in the handler or you'll get an error on return from the handler (rule is that 'all out parameters must be assigned on exit').

With the default init semantics, the only time you get an unassigned slot _apart from out parameters_ is for types which have no implicit default value which is a little inconsistent now. One option would be to make it so that 'out' parameters are initialized on entry - with this option you should never get an error related to an unassigned slot for any type which has a default value for any code generated by the compiler.

The only 'drawback' I see to this is that (with defaulting semantics) you won't get an error if you fail to set an out parameter in the handler (since it will, by default be initialized). However, that perhaps isn't so different from:

```
   handler foo() returns String
       variable tResult as String
       return tResult
   end handler
```

i.e. If defaulting is to be present, then it probably should be uniform to all slots.

peter-b commented on libscript/src/script-module.cpp:
```diff
@@ -383,6 +383,11 @@ bool MCScriptValidateModule(MCScriptModuleRef self)
 	                    }
 	                    break;
+                    case kMCScriptBytecodeOpReset:
+                        // create <reg_1>, ..., <reg_n>
```
"create" -> "reset"

peter-b commented:
@runrevmark Wow, this was fast work!  A few comments:
- It might be worth mentioning in the commit log for the "cleanup" commit that it rips out the current implementation of context variables.
- It'll certainly be nice to refactor the VM tests around `bytecode` blocks (so that we can test the VM directly, rather than indirectly)! Your current approach of moving the compiler tests into a `tests/lcb/compiler` tree is probably best for now. We do need a "proper" compiler testsuite at some point!
- I enjoyed the way that the `EnterScope` and `LeaveScope` actions were already defined in the compiler.
- For consistency, I think it does seem to make sense to auto-initialize out-parameters. _However_, not setting a non-`optional` out-parameter is probably a good indication that your program is Doing It Wrong. So some sort of compile-time warning would be nice if a function contains no reachable code that sets each of its out parameters.
- For now, I think it's okay to leave modules constructed by the VM and to address the problem of module instance constructors/destructors for another time.

runrevmark committed:
[[ CleanupVM ]] Clean up the VM code to allow easier changes.

This patch:
  - removes the unused context variable implementation
  - refactors register fetch and stores to ensure they go through
    (Checked)Fetch/StoreFrom/ToRegisterInFrame
  - gets rid of variable shadowing warnings in various for loops
  - gets rid of integer conversion warnings from several functions

runrevmark committed:
[[ LexicalScoping ]] Make local variables lexically scoped

This patch makes local variables lexically scoped. Specifically,
any 'variable' definitions inside within a block are only visible
after the point of definition up until the end of the current
block.

runrevmark committed:
[[ LocalVariableInit ]] Make compiler responsible for initializing locals

This patch moves the 'defaulting' action on local variables from an
implicit behavior to an explicit one by adding a 'reset' bytecode op
and adds defaulting of out parameters.

The compiler generates 'reset' opcodes for all variable definitions
at the point of definition.

runrevmark committed:
[[ Extensions ]] Fix variable usage in the gradient ramp editor

This patch ensures that the gradient ramp editor's code works
with lexically scoped variables.

runrevmark committed:
[[ NilAsUnassigned ]] Make the nil pointer mean unassigned

This patch changes the VM to make the nil pointer mean that a
slot is currently unassigned. This allows there to be a distinction
between kMCNull (nothing) and slots which have no value.

peter-b commented:
@runrevmark Could you please expand the release note with some more info about the implications of the change to lexical scope for variables?  I think it would be a good idea to also mention that `out` parameters are now default initialised.

peter-b commented:
@runrevmark I think there probably also need to be compiler tests for lexical scoping.

peter-b commented:
This is awesome 🎉 

@livecode-vulcan review ok 9a81c09

livecode-vulcan commented:
💙 review by @peter-b ok 9a81c09558cfc5177b62c68f77d5ef7c7824c272

livecode-vulcan commented:
😎 test success 9a81c09558cfc5177b62c68f77d5ef7c7824c272
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2602)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2612)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2597)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2613)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2619)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2625)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2637)
